File: onebusaway-android/src/main/java/org/onebusaway/android/ui/SituationDialog.java
/*
 * Copyright (C) 2012 Paul Watts (paulcwatts@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.onebusaway.android.ui;

import org.onebusaway.android.R;
import org.onebusaway.android.io.elements.ObaSituation;
import org.onebusaway.android.provider.ObaContract;
import org.onebusaway.android.util.UIUtils;

import android.app.Dialog;
import android.content.ContentValues;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.support.design.widget.Snackbar;
import android.support.v4.app.FragmentActivity;
import android.support.v7.app.AlertDialog;
import android.text.TextUtils;
import android.text.style.ClickableSpan;
import android.view.View;
import android.widget.TextView;

/**
 * Displays service alerts (i.e., situations) is a dialog
 */
public class SituationDialog {

    public static final String TAG = "SituationFragment";

    interface Listener {

        /**
         * Called when this dialog is dismissed
         *
         * @param isAlertHidden true if the service alert was hidden by the user, false if it was
         *                      not
         */
        void onDismiss(boolean isAlertHidden);

        /**
         * Called when the user taps the "Undo" snackbar for hiding an alert
         */
        void onUndo();
    }

    /**
     * Helper method to show this dialog
     */
    static void showDialog(final FragmentActivity activity, final ObaSituation situation,
            final Listener listener) {
        AlertDialog.Builder builder = new AlertDialog.Builder(activity)
                .setView(R.layout.situation)
                .setPositiveButton(R.string.hide, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int which) {
                        Dialog dialog = (Dialog) dialogInterface;

                        // Update the database to indicate that this alert has been hidden
                        ObaContract.ServiceAlerts
                                .insertOrUpdate(situation.getId(), new ContentValues(), false,
                                        true);

                        // Show the UNDO snackbar
                        Snackbar.make(activity.findViewById(R.id.fragment_arrivals_list),
                                R.string.alert_hidden_snackbar_text, Snackbar.LENGTH_SHORT)
                                .setAction(R.string.alert_hidden_snackbar_action,
                                        new View.OnClickListener() {
                                            @Override
                                            public void onClick(View v) {
                                                ObaContract.ServiceAlerts
                                                        .insertOrUpdate(situation.getId(),
                                                                new ContentValues(), false,
                                                                false);
                                                if (listener != null) {
                                                    listener.onUndo();
                                                }
                                            }
                                        }).show();
                        dialog.dismiss();
                        if (listener != null) {
                            listener.onDismiss(true);
                        }
                    }
                })
                .setNegativeButton(R.string.close, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                        if (listener != null) {
                            listener.onDismiss(false);
                        }
                    }
                });

        final android.support.v7.app.AlertDialog dialog = builder.create();

        dialog.show();

        // Set the title, description, and URL (if provided)
        TextView title = (TextView) dialog.findViewById(R.id.alert_title);
        title.setText(situation.getSummary());

        TextView desc = (TextView) dialog.findViewById(R.id.alert_description);
        desc.setText(situation.getDescription());

        TextView urlView = (TextView) dialog.findViewById(R.id.alert_url);

        // Remove any previous clickable spans just to be safe
        UIUtils.removeAllClickableSpans(urlView);

        if (!TextUtils.isEmpty(situation.getUrl())) {
            urlView.setVisibility(View.VISIBLE);

            ClickableSpan urlClick = new ClickableSpan() {
                public void onClick(View v) {
                    activity.startActivity(
                            new Intent(Intent.ACTION_VIEW, Uri.parse(situation.getUrl())));
                }
            };
            UIUtils.setClickableSpan(urlView, urlClick);
        } else {
            urlView.setVisibility(View.GONE);
        }

        // Update the database to indicate that this alert has been read
        ObaContract.ServiceAlerts
                .insertOrUpdate(situation.getId(), new ContentValues(), true, null);

        // TODO - analytics
        //ObaAnalytics.reportFragmentStart(dialog);
    }
}

File: onebusaway-android/src/main/java/org/onebusaway/android/ui/arrivals/dialogs/SituationDialog.kt
/*
 * Copyright (C) 2012-2026 Paul Watts (paulcwatts@gmail.com), Open Transit Software Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.onebusaway.android.ui.arrivals.dialogs

import android.content.Intent
import android.net.Uri
import android.text.Html
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.method.LinkMovementMethod
import android.text.style.ClickableSpan
import android.text.util.Linkify
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import org.onebusaway.android.R
import org.onebusaway.android.ui.arrivals.AlertDetails

/**
 * Shows a service alert (situation) in a dialog and marks it read. The buttons mirror the legacy
 * SituationDialogFragment: Hide flags this alert hidden (with an Undo snackbar), Hide All hides
 * every current alert and sets the hide-new-alerts preference, Close just dismisses.
 *
 * The persistence of read/hidden state is delegated to the host via callbacks so this dialog stays
 * pure UI (no ContentProvider/preference access) — the host wires them to the ViewModel/repository.
 *
 * @param onMarkRead persist that the alert was read (invoked when the dialog opens)
 * @param onHide persist that this alert is hidden (the Hide button)
 * @param onUnhide persist that this alert is no longer hidden, and refresh (the Undo action)
 * @param onHideAll hide every current alert and suppress new ones (the Hide All button)
 * @param onDismiss called when the dialog closes; true when the alert was hidden by the user
 * @param showUndoSnackbar shows a snackbar (optionally with an undo action). The host supplies this so
 *   the dialog doesn't reach for a specific View — the standalone activity anchors it to its arrivals
 *   root, while the Compose hosts (home sheet, NavHost destination) drive a Compose `SnackbarHost`.
 */
fun showSituationDialog(
    activity: AppCompatActivity,
    alert: AlertDetails,
    onMarkRead: () -> Unit,
    onHide: () -> Unit,
    onUnhide: () -> Unit,
    onHideAll: () -> Unit,
    onDismiss: (isAlertHidden: Boolean) -> Unit,
    showUndoSnackbar: (messageRes: Int, actionRes: Int?, onAction: (() -> Unit)?) -> Unit
) {
    val dialog = MaterialAlertDialogBuilder(activity)
        .setTitle(alert.summary)
        .setMessage(buildMessage(activity, alert))
        .setPositiveButton(R.string.hide) { d, _ ->
            onHide()
            showUndoSnackbar(
                R.string.alert_hidden_snackbar_text,
                R.string.alert_hidden_snackbar_action
            ) {
                onUnhide()
            }
            d.dismiss()
            onDismiss(true)
        }
        .setNeutralButton(R.string.hide_all) { d, _ ->
            onHideAll()
            showUndoSnackbar(R.string.all_alert_hidden_snackbar_text, null, null)
            d.dismiss()
            onDismiss(true)
        }
        .setNegativeButton(R.string.close) { d, _ ->
            d.dismiss()
            onDismiss(false)
        }
        .show()

    // Make the description's links (both linkified URLs and "More details") tappable
    dialog.findViewById<TextView>(android.R.id.message)?.movementMethod =
        LinkMovementMethod.getInstance()

    onMarkRead()
}

/**
 * The dialog body: the alert description (HTML rendered, bare URLs/phones/emails linkified, like
 * the legacy autoLink TextView) plus a "More details" link when the situation has a URL.
 */
private fun buildMessage(activity: AppCompatActivity, alert: AlertDetails): CharSequence {
    val description = alert.description.orEmpty()
    val message = SpannableStringBuilder(
        if (description.isNotEmpty()) {
            @Suppress("DEPRECATION")
            Html.fromHtml(description.replace(Regex("\\r\\n|\\r|\\n"), "<br>"))
        } else {
            activity.getString(R.string.no_description_available)
        }
    )
    @Suppress("DEPRECATION")
    Linkify.addLinks(message, Linkify.ALL)

    val url = alert.url
    if (!url.isNullOrEmpty()) {
        message.append("\n\n")
        val start = message.length
        message.append(activity.getString(R.string.situation_more_details))
        message.setSpan(
            object : ClickableSpan() {
                override fun onClick(widget: View) {
                    activity.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(url)))
                }
            },
            start, message.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
        )
    }
    return message
}
